### PR TITLE
puppet_utils: Do not force lang for cmd

### LINKED ISSUE #8000: Puppet fails with `invalid byte sequence in US-ASCII`
### Summary

When executing the puppet agent via ansible on a remote host, the command fails with `invalid byte sequence in US-ASCII`.
Executing the puppet agent on the remote host via a shell the command successfully ends with return code 0.

It looks due to shell environment used by ansible to execute the command the shell is not utf-8 compatible

### Issue Type

Bug Report

### Component Name

puppet

### Ansible Version

```console (paste below)
$ ansible --version
ansible [core 2.14.6]
```

### Community.general Version

```console (paste below)
$ ansible-galaxy collection list community.general

community.general 8.3.0
```

### Configuration

```console (paste below)
$ ansible-config dump --only-changed
CALLBACKS_ENABLED = ['ansible.posix.timer', 'ansible.posix.profile_tasks', 'ansible.posix.profile_roles']
COLLECTIONS_PATHS = ['/home/niklas/Documents/inovex/cloud/inovex-cloud-ansible/external/collections']
DEFAULT_FORKS = 30
DEFAULT_STDOUT_CALLBACK = community.general.diy
ENABLE_TASK_DEBUGGER = True
HOST_KEY_CHECKING = False
RETRY_FILES_ENABLED = False
```

### OS / Environment

Local: Ubuntu 23.10
Remote: Ubuntu 22.04

### Steps to Reproduce

Because of the dependency to puppet and the puppet code I cannot give an example to reproduce the bug. The latest puppet 7 version is used as a server and agent

This is the ansible code used to invoke the module
```yaml
- name: puppet run
  community.general.puppet:
    puppetmaster: "{{ ansible_fqdn }}"
    timeout: 60m
    verbose: true
    debug: true
    logdest: all
  register: puppet_agent_result
  until: puppet_agent_result.rc == 0
```

### Expected Results

I expect that the puppet run finishes with return code 0 or 2 (puppet specifies this as success but changes were executed).

### Actual Results

From the puppet logs

```console (paste below)
Error: Could not prefetch <module>: invalid byte sequence in US-ASCII
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by Knalltuete5000**
##### SUMMARY
The shell environment used forces by default the lang to be "C". This can result that puppet fails with the error message `invalid byte sequence in US-ASCII`.

Fixes #8000

##### ISSUE TYPE
Bugfix Pull Request

##### COMPONENT NAME
puppet

### CODE REVIEW

File: changelogs/fragments/puppet_lang_force.yml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+  - puppet - do not force the lang environment to 'C.UTF-8' instead of the default 'C' (https://github.com/ansible-collections/community.general/pull/8001)
```
Comment (felixfontein) [inline changelogs/fragments/puppet_lang_force.yml:2]:
```suggestion
  - puppet - do not force the lang environment to ``C.UTF-8`` instead of the default ``C.UTF-8`` (https://github.com/ansible-collections/community.general/pull/8001).
```

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution!

Do you mean `C` or `C.UTF-8`? The changelog fragment mentions something else than the code.

File: changelogs/fragments/puppet_lang_force.yml (hunk 1 above)
Comment (Knalltuete5000) [reply changelogs/fragments/puppet_lang_force.yml]: You are absolutely right. Adjusted the message.

File: changelogs/fragments/puppet_lang_force.yml [hunk 2]
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+  - puppet - do not force the lang environment to 'C'. Instead use 'C.UTF-8' (https://github.com/ansible-collections/community.general/pull/8001)
```
Comment (felixfontein) [inline changelogs/fragments/puppet_lang_force.yml:2]:
```suggestion
  - puppet - do not force the lang environment to ``C``. Instead use ``C.UTF-8`` (https://github.com/ansible-collections/community.general/pull/8001).
```

**Review by felixfontein — COMMENTED** [review]
Please note that this change is potentially problematic since `C.UTF-8` is not always available, which can cause the module to stop working for some users. (https://unix.stackexchange.com/questions/597962/how-widespread-is-the-c-utf-8-locale)

Maybe it's better to make this configurable, with `C.UTF-8` as a new default, so that at least users can explicitly provide another locale to fix when this breaks for them?

**Comment by felixfontein** [comment]
Also please note that you have to adjust the tests. They currently fail.

**Comment by Knalltuete5000** [comment]
> Please note that this change is potentially problematic since `C.UTF-8` is not always available, which can cause the module to stop working for some users. (https://unix.stackexchange.com/questions/597962/how-widespread-is-the-c-utf-8-locale)
> 
> 
> 
> Maybe it's better to make this configurable, with `C.UTF-8` as a new default, so that at least users can explicitly provide another locale to fix when this breaks for them?

This would be one solution I had in mind. Another one would be to pass in `None` as the lang. From my understanding this would then default to the system language, am I right?
What is the preferred approach for this fix

**Comment by felixfontein** [comment]
> This would be one solution I had in mind. Another one would be to pass in `None` as the lang. From my understanding this would then default to the system language, am I right?

Yes. But that's also a potentially breaking change, since it could be that the output changes depending on the system language.

I think providing the locale as a parameter is probably the best way to fix this, or alternatively writing code which detects whether C.UTF-8 is available.

**Comment by felixfontein** [comment]
(Assuming of course there isn't a way to tell puppet to not emit non-ASCII...)

**Comment by Knalltuete5000** [comment]
I have added a option via `environment_lang` to set the lang to force to. It will default to `C` which is the original value puppet/the cmdrunner was forced to

**Comment by Knalltuete5000** [comment]
Thnaks for the hit. I will look into it

**Comment by frittentheke** [comment]
> We had other cases of modules for which the default encoding was a moving target. Please take a look at other modules using `CmdRunner` for ideas.

@russoz  I know this the beyond this very issue here, but please let me say that having an option `force_lang` that actually is not an option which allows to optionally force a certain lang, but is actually actively forcing the lang to be "C" (https://github.com/ansible-collections/community.general/blob/21404851485e2c19f3daac4819fbbece7ada2e94/plugins/module_utils/cmd_runner.py#L194) if you don't force it to something else is a little odd and unexpected.

Be it as it may, I believe the approach used for the Ansible Galaxy install that @Knalltuete5000 mentioned (https://github.com/ansible-collections/community.general/blob/21404851485e2c19f3daac4819fbbece7ada2e94/plugins/modules/ansible_galaxy_install.py#L254) could actually be generalized inside the CmdRunner. So why not use `C.UTF-8` if available and only fallback to "C". All in case there is no explicit override done via `force_lang` to still allow forcing it.

**Comment by russoz** [comment]
> > We had other cases of modules for which the default encoding was a moving target. Please take a look at other modules using `CmdRunner` for ideas.
> 
> I have looked through the code, where the `CmdRunner` is used. I have found some examples but they will not solve the problem. Either the parameter `environ_update` is set, in near all cases, to `{'LANG': 'C', 'LC_ALL': 'C', 'LC_MESSAGES': 'C'}` or in the `ansible_galaxy_install.py` the `force_lang` is used. Here the `LANG` is set to `C.UTF-8`, if the command fails a retry is in place with the `LANG` set to `en_US.UTF-8` .
> 
> So what is the best way to solve the problem. With `LANG C` the problem still exist, to make it adjustable or to force it to an other `LANG` are the best options in my opinion.
> 
> I am open for any possible solutions how to solve the problem

IMHO, the Best Way™ would be to find a way (a parameter or an env variable) to disable the fancy output in `puppet`. That's what I did in a couple of other modules. When that is achievable, the encoding problem disappears as the output's encoding becomes 100% compatible with `C`.

Failing that, the only way I can see for the time being is for the code to perform tests with the command, using encoding A, if it fails, then try B, then C, etc... until hitting one that works. Something similar to what @frittentheke has suggested - although...

... he suggests making that a generic feature of the `CmdRunner` code, which is kinda tricky, as it would require a "no-op" command and that depends entirely on which executable is being used in that particular module. Many (most?) executables won't be idempotent, so running the _real_ command twice or more is not an option for such cases, and that complicates a generic solution.

**Comment by frittentheke** [comment]
@russoz @felixfontein  I'd really like for this issue / PR to move towards a fix and merge and not risk this to become stale.

1)
That said .... Puppet agent does support different locales, see https://www.puppet.com/docs/puppet/7/config_about_settings.html#configuring_locale_settings on how that is done.

If I may quote the relevant section ...
> You can use environment variables to set your locale for processes started on the command line. For most Linux distributions, set the LANG variable to your preferred locale, and the LANGUAGE variable to an empty string. On SLES, also set the LC_ALL variable to an empty string.

I am still wondering why it does not seem to respect `C`. Looking at https://github.com/ansible-collections/community.general/blob/21404851485e2c19f3daac4819fbbece7ada2e94/plugins/module_utils/cmd_runner.py#L255 only `LANGUAGE` and `LC_ALL` are set, so LANG remains untouched / unset? Also looking through the code via https://github.com/search?q=org%3Apuppetlabs+LC_ALL&type=code it seems those two variables usually go together.

Maybe that could be yet another way to potentially fix the issue @Knalltuete5000 and likely @skhan28 reported?

2)
> ... he suggests making that a generic feature of the `CmdRunner` code, which is kinda tricky, as it would require a "no-op" command and that depends entirely on which executable is being used in that particular module. Many (most?) executables won't be idempotent, so running the _real_ command twice or more is not an option for such cases, and that complicates a generic solution.

No @russoz, this is not what I meant when referencing https://github.com/ansible-collections/community.general/blob/21404851485e2c19f3daac4819fbbece7ada2e94/plugins/modules/ansible_galaxy_install.py#L254
I suggested to actually implement some sort of auto-discovery to first determine if UTF-8 is available on the system and then use this for any `CmdRunner` invoked commands, no attempts ... if it's available use it. And only if UTF-8 was not available, fall back to `C`. To say this bluntly, it's 2024 and it might even be more helpful to use UTF-8 as that is likely what most users would use interactively.

3) But also the suggested change with the introduction of a variable seems very sensible. So if 1) was not a surprise success to really force `C` on the puppet agent, I'd rather have a switch to flip to make it work.





In any case, please let's fix this issue here in any way you are happy with. Not being able to randomly invoke the Puppet agent  in some cases is the worse condition to keep this matter in.

File: plugins/module_utils/puppet.py [hunk 3]
```diff
@@ -4,14 +4,14 @@
...
 from __future__ import absolute_import, division, print_function
+
 __metaclass__ = type
 
 
 import os
 
 from ansible_collections.community.general.plugins.module_utils.cmd_runner import CmdRunner, cmd_runner_fmt
 
-
```
Comment (felixfontein) [inline plugins/module_utils/puppet.py:14]: If possible please avoid unrelated changes in PRs.

**Comment by Knalltuete5000** [comment]
Thanks for coming back to the issue and the review.

I have added the recommended changes by @felixfontein and removed the unrelated changes.

**Comment by felixfontein** [comment]
This looks good to me. While I agree that an auto-detection would be better, this is a lot harder to get right, and we can still do it later (like do the auto-detection if `environment_lang=auto`, and eventually change the default from `C` to `auto`).

If nobody objects, I'll merge this in ~a week.

**Comment by felixfontein** [comment]
@Knalltuete5000 thanks for your contribution!
@frittentheke @russoz thanks for reviewing and discussing this!
